test/shared/utils.py

```diff
@@ -37,10 +50,17 @@ def cibuildwheel_run(project_path, env=None, add_env=None):
     if add_env is not None:
         env.update(add_env)
 
-    subprocess.check_call(
-        [sys.executable, '-m', 'cibuildwheel', project_path],
-        env=env,
-    )
+    with TemporaryDirectory() as output_dir:
+        try:
+            subprocess.check_call(
+                [sys.executable, '-m', 'cibuildwheel', '--output-dir', output_dir, project_path],
+                env=env,
+            )
+        except:
+            assert len(os.listdir(output_dir)) == 0
```

Comment:
Hmmm, interesting. Are we sure about this?

Reply:
I was hesitant about this and thought I'd get feedback.
While all tests are passing as of now, this check is not applicable in the general case.
Wheels are put in the output directory config per config on Windows/macOS or all at once on Linux.
The behavior is thus different depending on the platform. On Linux, you either get all or none of the wheels depending of a single error. On macOS/Windows, you get all wheels with tests successful (if applicable) up until the error.
Either we consider the linux behavior to be the right one and this check can be left as-is (requiring to add further tests later to ensure this behavior) or I can make this check optional for the one test that does use it (and we keep a different behavior depending on the platform).

Reply:
> The behavior is thus different depending on the platform.

Good point. I hadn't even considered that. But that's a good point.
I was more thinking in terms that it should be the test asserting (because asserting no wheels are built is part of the test-), rather than here in `utils`.

In the end, the difference between OSs doesn't matter that much, I'd say? The error is thrown anyway?

Reply:
There are two more solutions, though:

1. Catch all errors, and just return the list? Normally, the test should catch that the wrong wheels are built? If necessary, we can return a few things, like error status, output, and built wheels; or return built wheels and `None`/caught exception?

2. Throw a custom exception on catching this one, that "includes the list of built wheels", as a way of having an exception ánd return value?

Neither of these two solutions give me a very warm feeling, though.

Reply:
There's yet another solution which provides flexibility (but consistency-wise would be a bit lower).
Add an `output_dir=None` default arg to `cibuildwheel_run`, much like @Czaki proposed at one time in order to use `tmp_path` only on selected project before saying all tests shall use a temp folder.

My feeling is that all tests needs to run in a temp folder.
90% of them won't need to bother with an `output_dir` and just need the resulting wheel names and what's here just works. By adding `output_dir=None`, we allow selected tests that require access to `output_dir` to provide it.

Any thoughts ?

Reply:
> The behavior is thus different depending on the platform. On Linux, you either get all or none of the wheels depending of a single error.
No. If you build 64 an 32 bits, then it copy wheels in two packs.

Reply:
> No. If you build 64 an 32 bits, then it copy wheels in two packs.

Huh? We don't mount a directory into the docker image anymore, do we? Because CircleCI doesn't allow that?

Reply:
@mayeut 

> There's yet another solution which provides flexibility (but consistency-wise would be a bit lower).
> [...]
> Any thoughts ?

Hmmm, myes, interesting; I hadn't thought of that, indeed. I guess that's as good of a solution as the other two things I suggested (or even better), since it's less of a hack, and doesn't try to "return a value from a function that's throwing an exception". (Or maybe this is the sign that the abstraction of having `run_cibuildwheel` return a list of wheels is not completely right?)

Reply:
But you run separated docker containers for x86_64 and i686 wheels. 
and when container finish work then result is copied. So you will have two transactions.

Reply:
@Czaki Ooooh, yes, right. I thought you meant you get pairs of 32- and 64-bit wheels. But you get two separate batches of all 32- and all 64-bit wheels.

Reply:
> No. If you build 64 an 32 bits, then it copy wheels in two packs.

You're right. Doesn't change the fact that behavior is different though. Anyway, this is not the matter here.

> Or maybe this is the sign that the abstraction of having run_cibuildwheel return a list of wheels is not completely right?

I think the option doesn't hurt too much the abstraction.